Amend AudioSampleBufferConverter to support various encoder options

jyavenard opened the pull request:
#### 905c6be5b0c89abe3500a69ea1d1c6266d1f7451
<pre>
Amend AudioSampleBufferConverter to support various encoder options
<a href="https://bugs.webkit.org/show_bug.cgi?id=284355">https://bugs.webkit.org/show_bug.cgi?id=284355</a>
<a href="https://rdar.apple.com/141204354">rdar://141204354</a>

Reviewed by Youenn Fablet.

Add support for passing various encoder options to the AudioSampleBufferConverter&apos;s AudioConverter.

The size of the compressed packet is determined by setting the AudioStreamBasicDescription&apos;s mFramesPerPacket.

Tests to follow as part of the WebCodec&apos;s WPT. No change of behaviours at this stage.

* Source/WebCore/platform/audio/cocoa/AudioSampleBufferConverter.h:
(WebCore::AudioSampleBufferConverter::preSkip const): Add method to retrieve the preSkip value after creating the AudioConverter.
Will be used to create the Opus&apos;s codec description.
* Source/WebCore/platform/audio/cocoa/AudioSampleBufferConverter.mm:
(WebCore::AudioSampleBufferConverter::initAudioConverterForSourceFormatDescription):

Canonical link: <a href="https://commits.webkit.org/287719@main">https://commits.webkit.org/287719@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/21b6b01b27a332411b2885cca5c94b74b0da7267

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80621 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59628 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34555 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85145 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/31603 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82732 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68689 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7934 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/62970 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/31603 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83690 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/53090 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73393 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43275 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50378 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27552 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30062 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/71524 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28075 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86578 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7847 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/5533 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/71263 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8022 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69229 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70503 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17570 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14516 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13467 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7809 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/13328 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7648 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/11167 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9453 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/audio/cocoa/AudioSampleBufferConverter.mm:
```diff
@@ -172,11 +172,21 @@
 
-    UInt32 size = sizeof(m_destinationFormat);
-    if (auto error = PAL::AudioFormatGetProperty(kAudioFormatProperty_FormatInfo, 0, NULL, &size, &m_destinationFormat)) {
+    if (auto error = [&](auto& destinationFormat) {
```
That is not really easy to read.
How about adding a static helper routine instead?

youennf commented on Source/WebCore/platform/audio/cocoa/AudioSampleBufferConverter.mm:
```diff
@@ -172,11 +172,21 @@
+            result = PAL::AudioFormatGetProperty(kAudioFormatProperty_FormatInfo, 0, NULL, &size, &destinationFormat);
+            if (originalDestinationFormat.mFramesPerPacket) {
+                // Adjust mFramesPerPacket to match new sampling rate
+                destinationFormat.mFramesPerPacket = originalDestinationFormat.mFramesPerPacket / originalDestinationFormat.mSampleRate * destinationFormat.mSampleRate;
```
Why changing destinationFormat if result is false?

jyavenard replied:
it's a thrown away value if error, I thought it didn't matter.

youennf commented on Source/WebCore/platform/audio/cocoa/AudioSampleBufferConverter.mm:
```diff
@@ -250,7 +259,29 @@
+        if (m_options.bitrateMode) {
+            UInt32 bitrateMode = *m_options.bitrateMode == BitrateMode::Variable ? kAudioCodecBitRateControlMode_Variable : kAudioCodecBitRateControlMode_Constant;
+            if (auto error = PAL::AudioConverterSetProperty(m_converter, kAudioCodecPropertyBitRateControlMode, sizeof(bitrateMode), &bitrateMode))
+                RELEASE_LOG_ERROR(MediaStream, "AudioSampleBufferConverter setting kAudioCodecPropertyBitRateControlMode failed with %d", static_cast<int>(error));
```
I would use RELEASE_LOG_ERROR_IF here.
Ditto below

jyavenard replied:
hmmm, that _IF is redundant though, we are there because error isn't null. Maybe it's better to remove all the other RELEASE_LOG_ERROR_IF instead

jyavenard replied:
and using RELEASE_LOG_ERROR_IF and running the command there, wouldn't let us print the error value in the message.
So I'll leave it as-is.

jyavenard committed:
Amend AudioSampleBufferConverter to support various encoder options
https://bugs.webkit.org/show_bug.cgi?id=284355
rdar://141204354

Reviewed by Youenn Fablet.

Add support for passing various encoder options to the AudioSampleBufferConverter's AudioConverter.

The size of the compressed packet is determined by setting the AudioStreamBasicDescription's mFramesPerPacket.

Tests to follow as part of the WebCodec's WPT. No change of behaviours at this stage.

* Source/WebCore/platform/audio/cocoa/AudioSampleBufferConverter.h:
(WebCore::AudioSampleBufferConverter::preSkip const): Add method to retrieve the preSkip value after creating the AudioConverter.
Will be used to create the Opus's codec description.
* Source/WebCore/platform/audio/cocoa/AudioSampleBufferConverter.mm:
(WebCore::AudioSampleBufferConverter::initAudioConverterForSourceFormatDescription):

Canonical link: https://commits.webkit.org/287719@main
